In [2]:
!pip install pandas

In [4]:
import pandas as pd
df= pd.read_csv("creditcard.csv")

In [6]:
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [8]:
df.shape

(284807, 31)

In [12]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [14]:
df.dtypes

Time      float64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [16]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [18]:
df.duplicated().sum()

1081

In [20]:
df= df.drop_duplicates()

In [22]:
df.duplicated().sum()

0

In [28]:
df.to_csv("creditcard_clean.csv", index=False)

In [32]:
df = pd.read_csv("data/creditcard_clean.csv")

In [34]:
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
283721  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
283722  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
283723  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
283724  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
283725  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
283721 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
283722  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
283723  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
283724  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
283725 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
283721  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
283722  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
283723 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
283724 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
283725  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
283721      0  
283722      0  
283723      0  
283724      0  
283725      0  

[283726 rows x 31 columns]

In [36]:
df["Class"].value_counts(normalize=True)

Class
0    0.998333
1    0.001667
Name: proportion, dtype: float64

In [38]:
df["Class"].value_counts()

Class
0    283253
1       473
Name: count, dtype: int64

In [40]:
df.describe()

Time             V1             V2             V3  \
count  283726.000000  283726.000000  283726.000000  283726.000000   
mean    94811.077600       0.005917      -0.004135       0.001613   
std     47481.047891       1.948026       1.646703       1.508682   
min         0.000000     -56.407510     -72.715728     -48.325589   
25%     54204.750000      -0.915951      -0.600321      -0.889682   
50%     84692.500000       0.020384       0.063949       0.179963   
75%    139298.000000       1.316068       0.800283       1.026960   
max    172792.000000       2.454930      22.057729       9.382558   

                  V4             V5             V6             V7  \
count  283726.000000  283726.000000  283726.000000  283726.000000   
mean       -0.002966       0.001828      -0.001139       0.001801   
std         1.414184       1.377008       1.331931       1.227664   
min        -5.683171    -113.743307     -26.160506     -43.557242   
25%        -0.850134      -0.689830      -0.769031      -0.552509   
50%        -0.022248      -0.053468      -0.275168       0.040859   
75%         0.739647       0.612218       0.396792       0.570474   
max        16.875344      34.801666      73.301626     120.589494   

                  V8             V9  ...            V21            V22  \
count  283726.000000  283726.000000  ...  283726.000000  283726.000000   
mean       -0.000854      -0.001596  ...      -0.000371      -0.000015   
std         1.179054       1.095492  ...       0.723909       0.724550   
min       -73.216718     -13.434066  ...     -34.830382     -10.933144   
25%        -0.208828      -0.644221  ...      -0.228305      -0.542700   
50%         0.021898      -0.052596  ...      -0.029441       0.006675   
75%         0.325704       0.595977  ...       0.186194       0.528245   
max        20.007208      15.594995  ...      27.202839      10.503090   

                 V23            V24            V25            V26  \
count  283726.000000  283726.000000  283726.000000  283726.000000   
mean        0.000198       0.000214      -0.000232       0.000149   
std         0.623702       0.605627       0.521220       0.482053   
min       -44.807735      -2.836627     -10.295397      -2.604551   
25%        -0.161703      -0.354453      -0.317485      -0.326763   
50%        -0.011159       0.041016       0.016278      -0.052172   
75%         0.147748       0.439738       0.350667       0.240261   
max        22.528412       4.584549       7.519589       3.517346   

                 V27            V28         Amount          Class  
count  283726.000000  283726.000000  283726.000000  283726.000000  
mean        0.001763       0.000547      88.472687       0.001667  
std         0.395744       0.328027     250.399437       0.040796  
min       -22.565679     -15.430084       0.000000       0.000000  
25%        -0.070641      -0.052818       5.600000       0.000000  
50%         0.001479       0.011288      22.000000       0.000000  
75%         0.091208       0.078276      77.510000       0.000000  
max        31.612198      33.847808   25691.160000       1.000000  

[8 rows x 31 columns]

📊 EDA Notes – Credit Card Fraud Detection Dataset

The dataset contains 283,726 transactions after deduplication, with 30 numerical features (Time, V1–V28, Amount) and 1 binary target variable (Class).

All features are numeric, making the dataset directly suitable for classical machine learning models without categorical encoding.

No missing values were found across any features, confirming data completeness.

1,081 exact duplicate transactions were identified and removed to prevent data leakage and artificial inflation of model performance.

The target variable is extremely imbalanced, with approximately 0.1667% fraudulent transactions and 99.8333% legitimate transactions.

Due to the severe class imbalance, accuracy is not an appropriate evaluation metric; instead, ROC-AUC and Precision–Recall AUC will be used.

Features V1–V28 are PCA-transformed and already standardized, while Time and Amount require scaling before modeling.

The dataset reflects real-world credit card fraud patterns, where fraudulent transactions are rare and decision thresholds must be tuned carefully.

In [46]:
!ls -la

total 429816
drwxr-xr-x  12 rajtharun  staff        384 Feb 10 19:30 .
drwx------@ 36 rajtharun  staff       1152 Feb 10 18:48 ..
-rw-r--r--@  1 rajtharun  staff       6148 Feb 10 19:23 .DS_Store
-rw-r--r--@  1 rajtharun  staff        127 Feb 10 19:28 .gitignore
drwxr-xr-x@  5 rajtharun  staff        160 Feb 10 19:27 .ipynb_checkpoints
-rw-rw-r--@  1 rajtharun  staff  150828752 Sep 20  2019 creditcard.csv
-rw-r--r--@  1 rajtharun  staff   69155672 Feb 10 18:44 creditcard.csv.zip
drwxr-xr-x   3 rajtharun  staff         96 Feb 10 19:02 data
-rw-r--r--@  1 rajtharun  staff      49826 Feb 10 19:30 eda_credit_card_fraud.ipynb
-rw-r--r--@  1 rajtharun  staff       7616 Feb 10 18:48 Machine learning.docx
-rw-r--r--@  1 rajtharun  staff        939 Feb 10 19:26 README.md
-rw-r--r--@  1 rajtharun  staff          0 Feb 10 19:25 untitled.txt


In [48]:
!cat .gitignore


__pycache__/
*.pyc
.ipynb_checkpoints/

# Ignore raw dataset
creditcard.csv
creditcard.csv.zip

# Environment files
.env
.venv


In [50]:
with open(".gitignore", "w") as f:
    f.write(
        "__pycache__/\n"
        "*.pyc\n"
        ".ipynb_checkpoints/\n"
        ".DS_Store\n\n"
        "# datasets (do not push)\n"
        "creditcard.csv\n"
        "creditcard.csv.zip\n"
        "data/creditcard_clean.csv\n\n"
        "# environment\n"
        ".env\n"
        ".venv\n"
    )


In [52]:
!cat .gitignore


__pycache__/
*.pyc
.ipynb_checkpoints/
.DS_Store

# datasets (do not push)
creditcard.csv
creditcard.csv.zip
data/creditcard_clean.csv

# environment
.env
.venv


In [54]:
!rm -f untitled.txt


In [56]:
!git status


On branch main

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	../.DS_Store
	../.localized
	../30 Days Leetcode.docx
	../AI:ML/
	../ALL FOLDERS/
	../Company Details/
	../DSA-Handbook-for-Coding-Interviews-main/
	../EY/
	../Eclipse Installer.app/
	../Google Chrome.app/
	../Leet Code/
	../Offer_Tharun_SDI.pdf
	../Resumes/
	../SQLDeveloper.app/
	"../Screenshot 2025-12-03 at 2.31.34\342\200\257PM.png"
	"../Screenshot 2025-12-11 at 2.40.52\342\200\257AM.png"
	"../Screenshot 2026-02-02 at 6.43.01\342\200\257PM.png"
	"../Screenshot 2026-02-02 at 6.48.50\342\200\257PM.png"
	"../Screenshot 2026-02-09 at 5.03.35\342\200\257PM.png"
	../Tharun Kumar_SE .pdf
	../content_factory/
	./
	../react-practice folder/
	../react-practice/
	../~$Amazon.docx
	../~$arun Kumar Nagipogu_Resume_SDE 2.docx
	../~$cap.docx
	../~$iprocv.docx
	../~$ite doc.docx
	../~$lide 1.docx
	../~$lyan documentation.docx
	../~$ture Drop.docx

nothing added to commit but untracked 

In [58]:
!pwd


/Users/rajtharun/Desktop/fraud-detection


In [60]:
!rm -rf .git
!rm -rf .gitmodules
!rm -rf .gitignore.lock 2>/dev/null



In [62]:
!git init
!git branch -M main



Initialized empty Git repository in /Users/rajtharun/Desktop/fraud-detection/.git/


In [64]:
!git rev-parse --show-toplevel
!git status


/Users/rajtharun/Desktop/fraud-detection
On branch main

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.gitignore
	Machine learning.docx
	README.md
	eda_credit_card_fraud.ipynb

nothing added to commit but untracked files present (use "git add" to track)


In [66]:
with open(".gitignore", "w") as f:
    f.write(
        "__pycache__/\n"
        "*.pyc\n"
        ".ipynb_checkpoints/\n"
        ".DS_Store\n\n"
        "# datasets (do not push)\n"
        "creditcard.csv\n"
        "creditcard.csv.zip\n"
        "data/creditcard_clean.csv\n\n"
        "# environment\n"
        ".env\n"
        ".venv\n"
    )


In [68]:
!rm -f untitled.txt


In [70]:
!git status


On branch main

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.gitignore
	Machine learning.docx
	README.md
	eda_credit_card_fraud.ipynb

nothing added to commit but untracked files present (use "git add" to track)
